diff --git a/packages/vite/src/node/plugins/importAnalysisBuild.ts b/packages/vite/src/node/plugins/importAnalysisBuild.ts
--- a/packages/vite/src/node/plugins/importAnalysisBuild.ts
+++ b/packages/vite/src/node/plugins/importAnalysisBuild.ts
@@ -643,13 +643,12 @@
   return [
     plugin,
     perEnvironmentPlugin('native:import-analysis-build', (environment) => {
-      const preloadCode = getPreloadCode(
-        environment,
-        !!renderBuiltUrl,
-        isRelativeBase,
-      )
       return nativeBuildImportAnalysisPlugin({
-        preloadCode,
+        base: environment.config.base,
+        modulePreloadPolyfill: !!(
+          environment.config.build.modulePreload &&
+          environment.config.build.modulePreload.polyfill
+        ),
         insertPreload: getInsertPreload(environment),
         // this field looks redundant, put a dummy value for now
         optimizeModulePreloadRelativePaths: false,
